homeassistant/components/goodwe/number.py

```diff
@@ -0,0 +1,102 @@
+"""GoodWe PV inverter numeric settings entities."""
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from dataclasses import dataclass
+import logging
+
+from goodwe import Inverter, InverterError
+
+from homeassistant.components.number import NumberEntity, NumberEntityDescription
+from homeassistant.const import ENTITY_CATEGORY_CONFIG, PERCENTAGE, POWER_WATT
+from homeassistant.helpers.entity import DeviceInfo
+
+from .const import DOMAIN, KEY_DEVICE_INFO, KEY_INVERTER
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class GoodweNumberEntityDescription(NumberEntityDescription):
+    """Class describing Goodwe sensor entities."""
+
+    getter: Callable[[Inverter], Awaitable[int]] | None = None
+    setter: Callable[[Inverter, int], Awaitable[None]] | None = None
+
+
+NUMBERS = (
+    GoodweNumberEntityDescription(
+        key="grid_export_limit",
+        name="Grid export limit",
+        icon="mdi:transmission-tower",
+        entity_category=ENTITY_CATEGORY_CONFIG,
+        unit_of_measurement=POWER_WATT,
+        getter=lambda inv: inv.get_grid_export_limit(),
+        setter=lambda inv, val: inv.set_grid_export_limit(val),
+        step=100,
+        min_value=0,
+        max_value=10000,
+    ),
+    GoodweNumberEntityDescription(
+        key="battery_discharge_depth",
+        name="Depth of discharge (on-grid)",
+        icon="mdi:battery-arrow-down",
+        entity_category=ENTITY_CATEGORY_CONFIG,
+        unit_of_measurement=PERCENTAGE,
+        getter=lambda inv: inv.get_ongrid_battery_dod(),
+        setter=lambda inv, val: inv.set_ongrid_battery_dod(val),
+        step=1,
+        min_value=0,
+        max_value=99,
+    ),
+)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the inverter select entities from a config entry."""
+    inverter = hass.data[DOMAIN][config_entry.entry_id][KEY_INVERTER]
+    device_info = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE_INFO]
+
+    entities = []
+
+    for description in NUMBERS:
+        try:
+            current_value = await description.getter(inverter)
+            entities.append(
```

Comment:
Please only wrap the line that can raise in the try... except block.

Reply:
I'm not sure I understand.
The point is one (or more) of the parameters may not be supported by the inverter.
The idea here was if the inverter will raise the exception (i.e. not supported), the entity will not be added.
If the response is ok, the entity is added and that's why it is within the try...except block.

Reply:
Only the line that can raise should be within the try... except block. This is to make it clear what we expect to raise and also to not catch errors unintentionally. The rest of the logic should be put elsewhere, eg by using an `else:` block or by returning in the except block or some other flow control.

In this case I suggest putting a `continue` in the except block and adding the entity to the list below outside the except block.

Reply:
Done